<a href="https://colab.research.google.com/github/Pragna235/ML-Assignments/blob/main/Credit_Data_Analysis_Task_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importing Modules and Libraries

In [5]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
import sys

# Reading Dataset

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
dataset = pd.read_csv("/content/drive/MyDrive/Credit.csv")
dataset

Unnamed: 0   Income  Limit  Rating  Cards  Age  Education  Gender  \
0             1   14.891   3606     283      2   34         11    Male   
1             2  106.025   6645     483      3   82         15  Female   
2             3  104.593   7075     514      4   71         11    Male   
3             4  148.924   9504     681      3   36         11  Female   
4             5   55.882   4897     357      2   68         16    Male   
..          ...      ...    ...     ...    ...  ...        ...     ...   
395         396   12.096   4100     307      3   32         13    Male   
396         397   13.364   3838     296      5   65         17    Male   
397         398   57.872   4171     321      5   67         12  Female   
398         399   37.728   2525     192      1   44         13    Male   
399         400   18.701   5524     415      5   64          7  Female   

    Student Married         Ethnicity  Balance  Defaultee  
0        No     Yes         Caucasian      333          0  
1       Yes     Yes             Asian      903          0  
2        No      No             Asian      580          0  
3        No      No             Asian      964          0  
4        No     Yes         Caucasian      331          0  
..      ...     ...               ...      ...        ...  
395      No     Yes         Caucasian      560          0  
396      No      No  African American      480          0  
397      No     Yes         Caucasian      138          0  
398      No     Yes         Caucasian        0          0  
399      No      No             Asian      966          0  

[400 rows x 13 columns]

# Dataset Analysis

In [7]:
# Print first 5 rows of the dataset
dataset.head()

Unnamed: 0   Income  Limit  Rating  Cards  Age  Education  Gender Student  \
0           1   14.891   3606     283      2   34         11    Male      No   
1           2  106.025   6645     483      3   82         15  Female     Yes   
2           3  104.593   7075     514      4   71         11    Male      No   
3           4  148.924   9504     681      3   36         11  Female      No   
4           5   55.882   4897     357      2   68         16    Male      No   

  Married  Ethnicity  Balance  Defaultee  
0     Yes  Caucasian      333          0  
1     Yes      Asian      903          0  
2      No      Asian      580          0  
3      No      Asian      964          0  
4     Yes  Caucasian      331          0

In [8]:
# Print last 5 rows of the dataset
dataset.tail()

Unnamed: 0  Income  Limit  Rating  Cards  Age  Education  Gender Student  \
395         396  12.096   4100     307      3   32         13    Male      No   
396         397  13.364   3838     296      5   65         17    Male      No   
397         398  57.872   4171     321      5   67         12  Female      No   
398         399  37.728   2525     192      1   44         13    Male      No   
399         400  18.701   5524     415      5   64          7  Female      No   

    Married         Ethnicity  Balance  Defaultee  
395     Yes         Caucasian      560          0  
396      No  African American      480          0  
397     Yes         Caucasian      138          0  
398     Yes         Caucasian        0          0  
399      No             Asian      966          0

In [9]:
# Print the number of rows and columns in the dataset
dataset.shape

(400, 13)

# Check for missing and duplicate values..

In [11]:
dataset.isnull().sum()

Unnamed: 0    0
Income        0
Limit         0
Rating        0
Cards         0
Age           0
Education     0
Gender        0
Student       0
Married       0
Ethnicity     0
Balance       0
Defaultee     0
dtype: int64

There are no missing values in the dataset

In [13]:
dataset[dataset.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, Income, Limit, Rating, Cards, Age, Education, Gender, Student, Married, Ethnicity, Balance, Defaultee]
Index: []

There are no duplicate rows


# Conversion of Categorical Values into Numerical Values

In [14]:
dataset.describe()

Unnamed: 0      Income         Limit      Rating       Cards  \
count  400.000000  400.000000    400.000000  400.000000  400.000000   
mean   200.500000   45.218885   4735.600000  354.940000    2.957500   
std    115.614301   35.244273   2308.198848  154.724143    1.371275   
min      1.000000   10.354000    855.000000   93.000000    1.000000   
25%    100.750000   21.007250   3088.000000  247.250000    2.000000   
50%    200.500000   33.115500   4622.500000  344.000000    3.000000   
75%    300.250000   57.470750   5872.750000  437.250000    4.000000   
max    400.000000  186.634000  13913.000000  982.000000    9.000000   

              Age   Education      Balance   Defaultee  
count  400.000000  400.000000   400.000000  400.000000  
mean    55.667500   13.450000   520.015000    0.180000  
std     17.249807    3.125207   459.758877    0.384669  
min     23.000000    5.000000     0.000000    0.000000  
25%     41.750000   11.000000    68.750000    0.000000  
50%     56.000000   14.000000   459.500000    0.000000  
75%     70.000000   16.000000   863.000000    0.000000  
max     98.000000   20.000000  1999.000000    1.000000

In [15]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  400 non-null    int64  
 1   Income      400 non-null    float64
 2   Limit       400 non-null    int64  
 3   Rating      400 non-null    int64  
 4   Cards       400 non-null    int64  
 5   Age         400 non-null    int64  
 6   Education   400 non-null    int64  
 7   Gender      400 non-null    object 
 8   Student     400 non-null    object 
 9   Married     400 non-null    object 
 10  Ethnicity   400 non-null    object 
 11  Balance     400 non-null    int64  
 12  Defaultee   400 non-null    int64  
dtypes: float64(1), int64(8), object(4)
memory usage: 40.8+ KB


The object type features should be converted to numerical values
[Gender,Student,Married,Ethnicity]

In [16]:
dataset.groupby('Defaultee').size()

Defaultee
0    328
1     72
dtype: int64

So, there are two classes of the target variable. That means, we have to determine if it's belonged to class 0 or class 1.


In [17]:
dataset = dataset.drop('Unnamed: 0',axis=1)
dataset

Income  Limit  Rating  Cards  Age  Education  Gender Student Married  \
0     14.891   3606     283      2   34         11    Male      No     Yes   
1    106.025   6645     483      3   82         15  Female     Yes     Yes   
2    104.593   7075     514      4   71         11    Male      No      No   
3    148.924   9504     681      3   36         11  Female      No      No   
4     55.882   4897     357      2   68         16    Male      No     Yes   
..       ...    ...     ...    ...  ...        ...     ...     ...     ...   
395   12.096   4100     307      3   32         13    Male      No     Yes   
396   13.364   3838     296      5   65         17    Male      No      No   
397   57.872   4171     321      5   67         12  Female      No     Yes   
398   37.728   2525     192      1   44         13    Male      No     Yes   
399   18.701   5524     415      5   64          7  Female      No      No   

            Ethnicity  Balance  Defaultee  
0           Caucasian      333          0  
1               Asian      903          0  
2               Asian      580          0  
3               Asian      964          0  
4           Caucasian      331          0  
..                ...      ...        ...  
395         Caucasian      560          0  
396  African American      480          0  
397         Caucasian      138          0  
398         Caucasian        0          0  
399             Asian      966          0  

[400 rows x 12 columns]

# Label Encoding


In [18]:
from sklearn.preprocessing import LabelEncoder

label = LabelEncoder()
dataset['Gender'] = label.fit_transform(dataset['Gender'])
dataset['Student'] = label.fit_transform(dataset['Student'])
dataset['Married'] = label.fit_transform(dataset['Married'])
dataset['Ethnicity'] = label.fit_transform(dataset['Ethnicity'])

dataset

Income  Limit  Rating  Cards  Age  Education  Gender  Student  Married  \
0     14.891   3606     283      2   34         11       1        0        1   
1    106.025   6645     483      3   82         15       0        1        1   
2    104.593   7075     514      4   71         11       1        0        0   
3    148.924   9504     681      3   36         11       0        0        0   
4     55.882   4897     357      2   68         16       1        0        1   
..       ...    ...     ...    ...  ...        ...     ...      ...      ...   
395   12.096   4100     307      3   32         13       1        0        1   
396   13.364   3838     296      5   65         17       1        0        0   
397   57.872   4171     321      5   67         12       0        0        1   
398   37.728   2525     192      1   44         13       1        0        1   
399   18.701   5524     415      5   64          7       0        0        0   

     Ethnicity  Balance  Defaultee  
0            2      333          0  
1            1      903          0  
2            1      580          0  
3            1      964          0  
4            2      331          0  
..         ...      ...        ...  
395          2      560          0  
396          0      480          0  
397          2      138          0  
398          2        0          0  
399          1      966          0  

[400 rows x 12 columns]

In [19]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 12 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Income     400 non-null    float64
 1   Limit      400 non-null    int64  
 2   Rating     400 non-null    int64  
 3   Cards      400 non-null    int64  
 4   Age        400 non-null    int64  
 5   Education  400 non-null    int64  
 6   Gender     400 non-null    int64  
 7   Student    400 non-null    int64  
 8   Married    400 non-null    int64  
 9   Ethnicity  400 non-null    int64  
 10  Balance    400 non-null    int64  
 11  Defaultee  400 non-null    int64  
dtypes: float64(1), int64(11)
memory usage: 37.6 KB


In [20]:
dataset['Defaultee'].value_counts()

0    328
1     72
Name: Defaultee, dtype: int64

# Model Training

In [21]:
from sklearn.model_selection import train_test_split

array = dataset.values
x = array[:,:-1]
y = array[:,-1]

x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.3,random_state=42)

# Training and Testing the Model using Logistic Regression

In [22]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(x_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [24]:
y_pred = model.predict(x_test)
y_pred

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 1., 1., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1.,
       0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       1., 0., 0., 1., 1., 1., 1., 0., 0., 1., 0., 0., 1., 1., 0., 0., 1.,
       0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 1., 1., 0., 0., 0., 0., 0.,
       0., 1., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0.])

In [27]:
from sklearn.metrics import classification_report, accuracy_score,confusion_matrix

accuracy_score(y_test,y_pred)

0.95

In [28]:
confusion_matrix(y_test,y_pred)

array([[96,  5],
       [ 1, 18]])

In [30]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

         0.0       0.99      0.95      0.97       101
         1.0       0.78      0.95      0.86        19

    accuracy                           0.95       120
   macro avg       0.89      0.95      0.91       120
weighted avg       0.96      0.95      0.95       120



So, the accuracy obtained using Logistic Regression is 95%

# Training and Testing the Model using Support Vector Machine


In [31]:
from sklearn.svm import SVC

model = SVC()
model.fit(x_train,y_train)

SVC()

In [32]:
y_pred = model.predict(x_test)
y_pred

array([1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0.,
       0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 1., 1., 0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.,
       0., 1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0.])

In [33]:
accuracy_score(y_test,y_pred)

0.8833333333333333

In [34]:
confusion_matrix(y_test,y_pred)

array([[97,  4],
       [10,  9]])

In [36]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

         0.0       0.91      0.96      0.93       101
         1.0       0.69      0.47      0.56        19

    accuracy                           0.88       120
   macro avg       0.80      0.72      0.75       120
weighted avg       0.87      0.88      0.87       120



So, the accuracy obtained using Support Vector Machine is approximately 89%

# Therefore, Logistic Regression did a better job than SVM in classifying if a person is defaultee or not.